src/state.py

```diff
@@ -20,23 +75,107 @@ def __init__(self, student_id, subjects_and_classes, subject_targets, subject_gi
+        del self.subject_targets[subject_name]
+
 
     def add_subject_give_in(self, subject_name, class_number):
-        flag = False
-        for sgi in self.subject_give_ins:
-            if sgi[0] == subject_name:
-                sgi[1].append(class_number)
-                flag = True
-        if not flag: self.subject_give_ins.append([subject_name, [class_number]])
-
-    
\ No newline at end of file
+        self.subject_give_ins[subject_name].append(class_number)
+
+
+
+    def get_class_for_subject(self, subject):
+        if subject in self.subjects_and_classes :
+            return self.subjects_and_classes[subject]
+        return None
+
+
+    def set_class_for_subject(self, subject, subject_class):
+        if subject in self.subjects_and_classes:
+            self.subjects_and_classes[subject] = subject_class
+
+
+    def get_targets_for_subject(self, subject):
+        if subject in self.subject_targets:
+            return self.subject_targets[subject]
+        return None
+
+    def __str__(self):
+        return "\nID: " + str(self.student_id) + "\nClasses: " + str(self.subjects_and_classes) + "\nTarget Classes: " + str(self.subject_targets) + "\nGive ins" + str(self.subject_give_ins)
+         
+
+    def get_giveins_for_subject(self, subject):
+        if subject in self.subject_give_ins:
+            return self.subject_give_ins[subject]
+
+        return None
+
+    
+    # equality comparator overload (compares student ids)
+    def __eq__(self, x):
+        return self.student_id == x.student_id
+
+
+
+#TESTING
+
+#First Student
+subjects_and_classes1 = {"LPOO": 1,"TCOM": 2,"MDIS": 1,"PLOG": 1}
+#subject_targets1 = {"LPOO": [2,3],"TCOM": [3,4],"MDIS":[5]}
+subject_targets1 = {"MDIS": [2]}
+subject_give_ins1 = {"MDIS": [3],"PLOG": [4,5]}
+
+
+a = Student(201603820,subjects_and_classes1,subject_targets1,subject_give_ins1)
+
+
+#Second student
+subjects_and_classes2 = {"LPOO": 3,"TCOM": 3,"MDIS": 2,"PLOG": 4}
+# subject_targets2 = {"LPOO": [1,4], "TCOM": [2,5]}
+subject_targets2 = {}
+subject_give_ins2 = {"MDIS": [1,3],"PLOG": [1,5]}
+b = Student(201703820,subjects_and_classes2,subject_targets2,subject_give_ins2)
+
+#Third student
+subjects_and_classes2 = {"LPOO": 5,"TCOM": 5,"MDIS": 5,"PLOG": 5}
+subject_targets2 = {"MDIS": [1]}
+subject_give_ins2 = {}
+c = Student(201803820,subjects_and_classes2,subject_targets2,subject_give_ins2)
+
+#Creates the state
+state = State()
+
+#Add both students
+state.add_student(a)
+state.add_student(b)
+# state.add_student(c)
+
+
+#print("\nID: ",a.student_id,"\nClasses: ",a.subjects_and_classes,"\nTarget Classes: ",a.subject_targets)
+#print("\nID: ",b.student_id,"\nClasses",b.subjects_and_classes,"\nTarget Classes",b.subject_targets)
+
+print("First State")
+for student in state.students.values():
+    print(student)
+
+    
+
+
+
+for neighbor in state.generate_neighbour():
+    print("\nNew State")
+    for student in neighbor.students.values():
+        print(student)
```

Comment:
I'd say we're good to go, just delete the tests (or save them locally if you wish) so our master branch is clean.